umodbus/serial.py

```diff
@@ -58,18 +52,197 @@ def __init__(self,
+class CommonRTUFunctions(object):
+    """Common Functions for Modbus RTU servers"""
+
+    def _calculate_crc16(self, data: bytearray) -> bytes:
+        """
+        Calculates the CRC16.
+
+        :param      data:        The data
+        :type       data:        bytearray
+
+        :returns:   The crc 16.
+        :rtype:     bytes
+        """
+        crc = 0xFFFF
+
+        for char in data:
+            crc = (crc >> 8) ^ Const.CRC16_TABLE[((crc) ^ char) & 0xFF]
+
+        return struct.pack('<H', crc)
+
+    def _exit_read(self, response: bytearray) -> bool:
+        """
+        Return on modbus read error
+
+        :param      response:    The response
+        :type       response:    bytearray
+
+        :returns:   State of basic read response evaluation
+        :rtype:     bool
+        """
+        if response[1] >= Const.ERROR_BIAS:
+            if len(response) < Const.ERROR_RESP_LEN:
+                return False
+        elif (Const.READ_COILS <= response[1] <= Const.READ_INPUT_REGISTER):
+            expected_len = Const.RESPONSE_HDR_LENGTH + 1 + \
+                           response[2] + Const.CRC_LENGTH
+            if len(response) < expected_len:
+                return False
+        elif len(response) < Const.FIXED_RESP_LEN:
+            return False
+
+        return True
+
+    def _validate_resp_hdr(self,
+                           response: bytearray,
+                           slave_addr: int,
+                           function_code: int,
+                           count: bool) -> bytes:
+        """
+        Validate the response header.
+
+        :param      response:       The response
+        :type       response:       bytearray
+        :param      slave_addr:     The slave address
+        :type       slave_addr:     int
+        :param      function_code:  The function code
+        :type       function_code:  int
+        :param      count:          The count
+        :type       count:          bool
+
+        :returns:   Modbus response content
+        :rtype:     bytes
+        """
+        if len(response) == 0:
+            raise OSError('no data received from slave')
+
+        resp_crc = response[-Const.CRC_LENGTH:]
+        expected_crc = self._calculate_crc16(
+            response[0:len(response) - Const.CRC_LENGTH]
+        )
+
+        if ((resp_crc[0] is not expected_crc[0]) or
+                (resp_crc[1] is not expected_crc[1])):
+            raise OSError('invalid response CRC')
+
+        if (response[0] != slave_addr):
+            raise ValueError('wrong slave address')
+
+        if (response[1] == (function_code + Const.ERROR_BIAS)):
+            raise ValueError('slave returned exception code: {:d}'.
+                             format(response[2]))
+
+        hdr_length = (Const.RESPONSE_HDR_LENGTH + 1) if count else \
+            Const.RESPONSE_HDR_LENGTH
+
+        return response[hdr_length:len(response) - Const.CRC_LENGTH]
+
+    def _form_serial_pdu(self,
```

Comment:
please add missing doc string

Reply:
Thanks for pointing that out, I completely forgot I split off those functions earlier. Added docstrings